From 4c1d9fb1e8531e77933219660f8c1e2e78b53ea5 Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Sun, 2 Feb 2014 20:45:14 +0000
Subject: [PATCH] Fix DBCP-341 LocalXAConnectionFactory does not properly check
 if Xid is equal to currentXid when resuming which may result in an
 XAException. Patch by Ioannis Canellos

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/branches/DBCP_1_5_x_BRANCH@1563710 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                                       | 4 ++++
 .../apache/commons/dbcp/managed/LocalXAConnectionFactory.java | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 791f55310a..3f03728cfe 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -83,6 +83,10 @@ The <action> type attribute can be add,update,fix,remove.
         connection is consistent when the underlying connection is closed as a
         result of an error condition.
       </action>
+      <action dev="markt" issue="DBCP-341" type="fix" due-to="Ioannis Canellos">
+        LocalXAConnectionFactory does not properly check if Xid is equal to
+        currentXid when resuming which may result in an XAException.
+      </action>
     </release>
     <release version="1.4.1" date="TBD" description="TBD">
       <action dev="psteitz" issue="DBCP-367" type="fix" due-to="Ken Tatsushita">
diff --git a/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java b/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
index b8ed0e87ef..a68996ab6f 100644
--- a/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
+++ b/src/main/java/org/apache/commons/dbcp/managed/LocalXAConnectionFactory.java
@@ -136,7 +136,7 @@ public synchronized void start(Xid xid, int flag) throws XAException {
 
                 this.currentXid = xid;
             } else if (flag == XAResource.TMRESUME) {
-                if (xid != this.currentXid) {
+                if (!xid.equals(this.currentXid)) {
                     throw new XAException("Attempting to resume in different transaction: expected " + this.currentXid + ", but was " + xid);
                 }
             } else {
